File: tools/opensbi/platform/kendryte/k210/sysctl.c
/*
 * SPDX-License-Identifier:    Apache-2.0
 *
 * Copyright 2018 Canaan Inc.
 * Copyright (c) 2019 Western Digital Corporation or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <sbi/riscv_encoding.h>
#include <sbi/sbi_timer.h>
#include "sysctl.h"

volatile sysctl_t *const sysctl = (volatile sysctl_t *)SYSCTL_BASE_ADDR;

#define SYSCTRL_CLOCK_FREQ_IN0 (26000000UL)

static u32 sysctl_pll0_get_freq(void)
{
	u32 freq_in, nr, nf, od;

	freq_in = SYSCTRL_CLOCK_FREQ_IN0;
	nr	= sysctl->pll0.clkr0 + 1;
	nf	= sysctl->pll0.clkf0 + 1;
	od	= sysctl->pll0.clkod0 + 1;

	/*
	 * Get final PLL output freq
	 * FOUT = FIN / NR * NF / OD
	 *      = (FIN * NF) / (NR * OD)
	 */
	return ((u64)freq_in * (u64)nf) / ((u64)nr * (u64)od);
}

u32 sysctl_get_cpu_freq(void)
{
	int clock_source;

	clock_source = (int)sysctl->clk_sel0.aclk_sel;
	switch (clock_source) {
	case 0:
		return SYSCTRL_CLOCK_FREQ_IN0;
	case 1:
		return sysctl_pll0_get_freq() /
		       (2ULL << (int)sysctl->clk_sel0.aclk_divider_sel);
	default:
		return 0;
	}
}

int sysctl_clock_bus_en(sysctl_clock_t clock, uint8_t en)
{
	/*
	 * The timer is under APB0, to prevent apb0_clk_en1 and apb0_clk_en0
	 * on same register, we split it to peripheral and central two
	 * registers, to protect CPU close apb0 clock accidentally.
	 *
	 * The apb0_clk_en0 and apb0_clk_en1 have same function,
	 * one of them set, the APB0 clock enable.
	 */

	/* The APB clock should carefully disable */
	if(en)
	{
		switch(clock)
		{
			/*
			 * These peripheral devices are under APB0
			 * GPIO, UART1, UART2, UART3, SPI_SLAVE, I2S0, I2S1,
			 * I2S2, I2C0, I2C1, I2C2, FPIOA, SHA256, TIMER0,
			 * TIMER1, TIMER2
			 */
		case SYSCTL_CLOCK_GPIO:
		case SYSCTL_CLOCK_SPI2:
		case SYSCTL_CLOCK_I2S0:
		case SYSCTL_CLOCK_I2S1:
		case SYSCTL_CLOCK_I2S2:
		case SYSCTL_CLOCK_I2C0:
		case SYSCTL_CLOCK_I2C1:
		case SYSCTL_CLOCK_I2C2:
		case SYSCTL_CLOCK_UART1:
		case SYSCTL_CLOCK_UART2:
		case SYSCTL_CLOCK_UART3:
		case SYSCTL_CLOCK_FPIOA:
		case SYSCTL_CLOCK_TIMER0:
		case SYSCTL_CLOCK_TIMER1:
		case SYSCTL_CLOCK_TIMER2:
		case SYSCTL_CLOCK_SHA:
			sysctl->clk_en_cent.apb0_clk_en = en;
			break;

			/*
			 * These peripheral devices are under APB1
			 * WDT, AES, OTP, DVP, SYSCTL
			 */
		case SYSCTL_CLOCK_AES:
		case SYSCTL_CLOCK_WDT0:
		case SYSCTL_CLOCK_WDT1:
		case SYSCTL_CLOCK_OTP:
		case SYSCTL_CLOCK_RTC:
			sysctl->clk_en_cent.apb1_clk_en = en;
			break;

			/*
			 * These peripheral devices are under APB2
			 * SPI0, SPI1
			 */
		case SYSCTL_CLOCK_SPI0:
		case SYSCTL_CLOCK_SPI1:
			sysctl->clk_en_cent.apb2_clk_en = en;
			break;

		default:
			break;
		}
	}

	return 0;
}

int sysctl_clock_device_en(sysctl_clock_t clock, uint8_t en)
{
	switch(clock)
	{
		/*
		 * These devices are PLL
		 */
	case SYSCTL_CLOCK_PLL0:
		sysctl->pll0.pll_out_en0 = en;
		break;
	case SYSCTL_CLOCK_PLL1:
		sysctl->pll1.pll_out_en1 = en;
		break;
	case SYSCTL_CLOCK_PLL2:
		sysctl->pll2.pll_out_en2 = en;
		break;

		/*
		 * These devices are CPU, SRAM, APB bus, ROM, DMA, AI
		 */
	case SYSCTL_CLOCK_CPU:
		sysctl->clk_en_cent.cpu_clk_en = en;
		break;
	case SYSCTL_CLOCK_SRAM0:
		sysctl->clk_en_cent.sram0_clk_en = en;
		break;
	case SYSCTL_CLOCK_SRAM1:
		sysctl->clk_en_cent.sram1_clk_en = en;
		break;
	case SYSCTL_CLOCK_APB0:
		sysctl->clk_en_cent.apb0_clk_en = en;
		break;
	case SYSCTL_CLOCK_APB1:
		sysctl->clk_en_cent.apb1_clk_en = en;
		break;
	case SYSCTL_CLOCK_APB2:
		sysctl->clk_en_cent.apb2_clk_en = en;
		break;
	case SYSCTL_CLOCK_ROM:
		sysctl->clk_en_peri.rom_clk_en = en;
		break;
	case SYSCTL_CLOCK_DMA:
		sysctl->clk_en_peri.dma_clk_en = en;
		break;
	case SYSCTL_CLOCK_AI:
		sysctl->clk_en_peri.ai_clk_en = en;
		break;
	case SYSCTL_CLOCK_DVP:
		sysctl->clk_en_peri.dvp_clk_en = en;
		break;
	case SYSCTL_CLOCK_FFT:
		sysctl->clk_en_peri.fft_clk_en = en;
		break;
	case SYSCTL_CLOCK_SPI3:
		sysctl->clk_en_peri.spi3_clk_en = en;
		break;

		/*
		 * These peripheral devices are under APB0
		 * GPIO, UART1, UART2, UART3, SPI_SLAVE, I2S0, I2S1,
		 * I2S2, I2C0, I2C1, I2C2, FPIOA, SHA256, TIMER0,
		 * TIMER1, TIMER2
		 */
	case SYSCTL_CLOCK_GPIO:
		sysctl->clk_en_peri.gpio_clk_en = en;
		break;
	case SYSCTL_CLOCK_SPI2:
		sysctl->clk_en_peri.spi2_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2S0:
		sysctl->clk_en_peri.i2s0_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2S1:
		sysctl->clk_en_peri.i2s1_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2S2:
		sysctl->clk_en_peri.i2s2_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2C0:
		sysctl->clk_en_peri.i2c0_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2C1:
		sysctl->clk_en_peri.i2c1_clk_en = en;
		break;
	case SYSCTL_CLOCK_I2C2:
		sysctl->clk_en_peri.i2c2_clk_en = en;
		break;
	case SYSCTL_CLOCK_UART1:
		sysctl->clk_en_peri.uart1_clk_en = en;
		break;
	case SYSCTL_CLOCK_UART2:
		sysctl->clk_en_peri.uart2_clk_en = en;
		break;
	case SYSCTL_CLOCK_UART3:
		sysctl->clk_en_peri.uart3_clk_en = en;
		break;
	case SYSCTL_CLOCK_FPIOA:
		sysctl->clk_en_peri.fpioa_clk_en = en;
		break;
	case SYSCTL_CLOCK_TIMER0:
		sysctl->clk_en_peri.timer0_clk_en = en;
		break;
	case SYSCTL_CLOCK_TIMER1:
		sysctl->clk_en_peri.timer1_clk_en = en;
		break;
	case SYSCTL_CLOCK_TIMER2:
		sysctl->clk_en_peri.timer2_clk_en = en;
		break;
	case SYSCTL_CLOCK_SHA:
		sysctl->clk_en_peri.sha_clk_en = en;
		break;

		/*
		 * These peripheral devices are under APB1
		 * WDT, AES, OTP, DVP, SYSCTL
		 */
	case SYSCTL_CLOCK_AES:
		sysctl->clk_en_peri.aes_clk_en = en;
		break;
	case SYSCTL_CLOCK_WDT0:
		sysctl->clk_en_peri.wdt0_clk_en = en;
		break;
	case SYSCTL_CLOCK_WDT1:
		sysctl->clk_en_peri.wdt1_clk_en = en;
		break;
	case SYSCTL_CLOCK_OTP:
		sysctl->clk_en_peri.otp_clk_en = en;
		break;
	case SYSCTL_CLOCK_RTC:
		sysctl->clk_en_peri.rtc_clk_en = en;
		break;

		/*
		 * These peripheral devices are under APB2
		 * SPI0, SPI1
		 */
	case SYSCTL_CLOCK_SPI0:
		sysctl->clk_en_peri.spi0_clk_en = en;
		break;
	case SYSCTL_CLOCK_SPI1:
		sysctl->clk_en_peri.spi1_clk_en = en;
		break;

	default:
		break;
	}

	return 0;
}

int sysctl_clock_enable(sysctl_clock_t clock)
{
	if(clock >= SYSCTL_CLOCK_MAX)
		return -1;
	sysctl_clock_bus_en(clock, 1);
	sysctl_clock_device_en(clock, 1);
	return 0;
}


static void sysctl_reset_ctl(sysctl_reset_t reset, uint8_t rst_value)
{
	switch(reset)
	{
	case SYSCTL_RESET_SOC:
		sysctl->soft_reset.soft_reset = rst_value;
		break;
	case SYSCTL_RESET_ROM:
		sysctl->peri_reset.rom_reset = rst_value;
		break;
	case SYSCTL_RESET_DMA:
		sysctl->peri_reset.dma_reset = rst_value;
		break;
	case SYSCTL_RESET_AI:
		sysctl->peri_reset.ai_reset = rst_value;
		break;
	case SYSCTL_RESET_DVP:
		sysctl->peri_reset.dvp_reset = rst_value;
		break;
	case SYSCTL_RESET_FFT:
		sysctl->peri_reset.fft_reset = rst_value;
		break;
	case SYSCTL_RESET_GPIO:
		sysctl->peri_reset.gpio_reset = rst_value;
		break;
	case SYSCTL_RESET_SPI0:
		sysctl->peri_reset.spi0_reset = rst_value;
		break;
	case SYSCTL_RESET_SPI1:
		sysctl->peri_reset.spi1_reset = rst_value;
		break;
	case SYSCTL_RESET_SPI2:
		sysctl->peri_reset.spi2_reset = rst_value;
		break;
	case SYSCTL_RESET_SPI3:
		sysctl->peri_reset.spi3_reset = rst_value;
		break;
	case SYSCTL_RESET_I2S0:
		sysctl->peri_reset.i2s0_reset = rst_value;
		break;
	case SYSCTL_RESET_I2S1:
		sysctl->peri_reset.i2s1_reset = rst_value;
		break;
	case SYSCTL_RESET_I2S2:
		sysctl->peri_reset.i2s2_reset = rst_value;
		break;
	case SYSCTL_RESET_I2C0:
		sysctl->peri_reset.i2c0_reset = rst_value;
		break;
	case SYSCTL_RESET_I2C1:
		sysctl->peri_reset.i2c1_reset = rst_value;
		break;
	case SYSCTL_RESET_I2C2:
		sysctl->peri_reset.i2c2_reset = rst_value;
		break;
	case SYSCTL_RESET_UART1:
		sysctl->peri_reset.uart1_reset = rst_value;
		break;
	case SYSCTL_RESET_UART2:
		sysctl->peri_reset.uart2_reset = rst_value;
		break;
	case SYSCTL_RESET_UART3:
		sysctl->peri_reset.uart3_reset = rst_value;
		break;
	case SYSCTL_RESET_AES:
		sysctl->peri_reset.aes_reset = rst_value;
		break;
	case SYSCTL_RESET_FPIOA:
		sysctl->peri_reset.fpioa_reset = rst_value;
		break;
	case SYSCTL_RESET_TIMER0:
		sysctl->peri_reset.timer0_reset = rst_value;
		break;
	case SYSCTL_RESET_TIMER1:
		sysctl->peri_reset.timer1_reset = rst_value;
		break;
	case SYSCTL_RESET_TIMER2:
		sysctl->peri_reset.timer2_reset = rst_value;
		break;
	case SYSCTL_RESET_WDT0:
		sysctl->peri_reset.wdt0_reset = rst_value;
		break;
	case SYSCTL_RESET_WDT1:
		sysctl->peri_reset.wdt1_reset = rst_value;
		break;
	case SYSCTL_RESET_SHA:
		sysctl->peri_reset.sha_reset = rst_value;
		break;
	case SYSCTL_RESET_RTC:
		sysctl->peri_reset.rtc_reset = rst_value;
		break;

	default:
		break;
	}
}

void usleep(unsigned long long usec) {
	unsigned long long future = read_time() + usec * 10;
	while (read_time() < future);
}

void sysctl_reset(sysctl_reset_t reset)
{
	sysctl_reset_ctl(reset, 1);
	usleep(10);
	sysctl_reset_ctl(reset, 0);
}

File: os/src/drivers/sysctl.rs
#![allow(dead_code)]

use bitflags::bitflags;

use crate::platform::SYSCTL_BASE_ADDR;
use crate::interrupt::timer::usleep;

const SYSCTL_CLK_EN_CENT_OFFSET: usize = 0x28;
bitflags! {
    pub struct SYSCTLClockEnableCent: u32 {
        const CPU_CLOCK_ENABLE = 1 << 0;
        const SRAM0_CLOCK_ENABLE = 1 << 1;
        const SRAM1_CLOCK_ENABLE = 1 << 2;
        const APB0_CLOCK_ENABLE = 1 << 3;
        const APB1_CLOCK_ENABLE = 1 << 4;
        const APB2_CLOCK_ENABLE = 1 << 5;
    }
}

impl SYSCTLClockEnableCent {
    fn get_ptr() -> *mut u32 {
        (SYSCTL_BASE_ADDR + SYSCTL_CLK_EN_CENT_OFFSET) as *mut u32
    }
    unsafe fn read() -> u32 {
        Self::get_ptr().read_volatile()
    }
    unsafe fn write(v: u32) {
        Self::get_ptr().write_volatile(v)
    }
    fn set_cpu_clock_enable(enable: bool) {
        let mut flags = Self::from_bits(
            unsafe { Self::read() }
        ).unwrap();
        flags.set(Self::CPU_CLOCK_ENABLE, enable);
        unsafe { Self::write(flags.bits) }
    }
    fn set_sram_clock_enable(sram_id: i32, enable: bool) {
        let mut flags = Self::from_bits(
            unsafe { Self::read() }
        ).unwrap();
        let bit = match sram_id {
            0 => Self::SRAM0_CLOCK_ENABLE,
            1 => Self::SRAM1_CLOCK_ENABLE,
            _ => { panic!("given SRAM id in set_sram_clock_enable() is not supported!") }
        };
        flags.set(bit, enable);
        unsafe { Self::write(flags.bits) }
    }
    fn set_apb_clock_enable(apb_id: i32, enable: bool) {
        let mut flags = Self::from_bits(
            unsafe { Self::read() }
        ).unwrap();
        let bit = match apb_id {
            0 => Self::APB0_CLOCK_ENABLE,
            1 => Self::APB1_CLOCK_ENABLE,
            2 => Self::APB2_CLOCK_ENABLE,
            _ => { panic!("given APB id in set_apb_block_enable() is not supported!") }
        };
        flags.set(bit, enable);
        unsafe { Self::write(flags.bits) }
    }
}

const SYSCTL_CLK_EN_PERI_OFFSET: usize = 0x2c;
bitflags! {
    pub struct SYSCTLClockEnablePeri: u32 {
        const ROM_CLOCK_ENABLE      = 1 << 0;
        const DMA_CLOCK_ENABLE      = 1 << 1;
        const AI_CLOCK_ENABLE       = 1 << 2;
        const DVP_CLOCK_ENABLE      = 1 << 3;
        const FFT_CLOCK_ENABLE      = 1 << 4;
        const GPIO0_CLOCK_ENABLE    = 1 << 5;
        const SPI0_CLOCK_ENABLE     = 1 << 6;
        const SPI1_CLOCK_ENABLE     = 1 << 7;
        const SPI2_CLOCK_ENABLE     = 1 << 8;
        const SPI3_CLOCK_ENABLE     = 1 << 9;
        const I2S0_CLOCK_ENABLE     = 1 << 10;
        const I2S1_CLOCK_ENABLE     = 1 << 11;
        const I2S2_CLOCK_ENABLE     = 1 << 12;
        const I2C0_CLOCK_ENABLE     = 1 << 13;
        const I2C1_CLOCK_ENABLE     = 1 << 14;
        const I2C2_CLOCK_ENABLE     = 1 << 15;
        const UART1_CLOCK_ENABLE    = 1 << 16;
        const UART2_CLOCK_ENABLE    = 1 << 17;
        const UART3_CLOCK_ENABLE    = 1 << 18;
        const AES_CLOCK_ENABLE      = 1 << 19;
        const FPIOA_CLOCK_ENABLE    = 1 << 20;
        const TIMER0_CLOCK_ENABLE   = 1 << 21;
        const TIMER1_CLOCK_ENABLE   = 1 << 22;
        const TIMER2_CLOCK_ENABLE   = 1 << 23;
        const WDT0_CLOCK_ENABLE     = 1 << 24;
        const WDT1_CLOCK_ENABLE     = 1 << 25;
        const SHA_CLOCK_ENABLE      = 1 << 26;
        const OTP_CLOCK_ENABLE      = 1 << 27;
        const RTC_CLOCK_ENABLE      = 1 << 29;
    }
}
impl SYSCTLClockEnablePeri {
    fn get_ptr() -> *mut u32 {
        (SYSCTL_BASE_ADDR + SYSCTL_CLK_EN_PERI_OFFSET) as *mut u32
    }
    unsafe fn read() -> u32 {
        Self::get_ptr().read_volatile()
    }
    unsafe fn write(v: u32) {
        Self::get_ptr().write_volatile(v)
    }
    fn set_uart_clock_enable(uart_id: i32, enable: bool) {
        let mut flags = Self::from_bits(
            unsafe { Self::read() }
        ).unwrap();
        let bit = match uart_id {
            1 => Self::UART1_CLOCK_ENABLE,
            2 => Self::UART2_CLOCK_ENABLE,
            3 => Self::UART3_CLOCK_ENABLE,
            _ => { panic!("invalid uart id!"); }
        };
        flags.set(bit, enable);
        unsafe { Self::write(flags.bits) }
    }
}

const SYSCTL_SOFT_RESET_OFFSET: usize = 0x30;
bitflags! {
    struct SYSCTLSoftReset: u32 {
        const SOFT_RESET = 1 << 0;
    }
}
impl SYSCTLSoftReset {
    fn get_ptr() -> *mut u32 {
        (SYSCTL_BASE_ADDR + SYSCTL_SOFT_RESET_OFFSET) as *mut u32
    }
    unsafe fn read() -> u32 {
        Self::get_ptr().read_volatile()
    }
    unsafe fn write(v: u32) {
        Self::get_ptr().write_volatile(v)
    }
    fn set_soft_reset(value: bool) {
        let mut flags = Self::from_bits( unsafe{ Self::read() }).unwrap();
        flags.set(Self::SOFT_RESET, value);
        unsafe { Self::write(flags.bits) }
    }
}

const SYSCTL_PERI_RESET_OFFSET: usize = 0x34;
bitflags! {
    struct SYSCTLPeriReset: u32 {
        const ROM_RESET     = 1 << 0;
        const DMA_RESET     = 1 << 1;
        const AI_RESET      = 1 << 2;
        const DVP_RESET     = 1 << 3;
        const FFT_RESET     = 1 << 4;
        const GPIO_RESET    = 1 << 5;
        const SPI0_RESET    = 1 << 6;
        const SPI1_RESET    = 1 << 7;
        const SPI2_RESET    = 1 << 8;
        const SPI3_RESET    = 1 << 9;
        const I2S0_RESET    = 1 << 10;
        const I2S1_RESET    = 1 << 11;
        const I2S2_RESET    = 1 << 12;
        const I2C0_RESET    = 1 << 13;
        const I2C1_RESET    = 1 << 14;
        const I2C2_RESET    = 1 << 15;
        const UART1_RESET   = 1 << 16;
        const UART2_RESET   = 1 << 17;
        const UART3_RESET   = 1 << 18;
        const AES_RESET     = 1 << 19;
        const FPIOA_RESET   = 1 << 20;
        const TIMER0_RESET  = 1 << 21;
        const TIMER1_RESET  = 1 << 22;
        const TIMER2_RESET  = 1 << 23;
        const WDT0_RESET    = 1 << 24;
        const WDT1_RESET    = 1 << 25;
        const SHA_RESET     = 1 << 26;
        const RTC_RESET     = 1 << 29;
    }
}
impl SYSCTLPeriReset {
    fn get_ptr() -> *mut u32 {
        (SYSCTL_BASE_ADDR + SYSCTL_PERI_RESET_OFFSET) as *mut u32
    }
    unsafe fn read() -> u32 {
        Self::get_ptr().read_volatile()
    }
    unsafe fn write(v: u32) {
        Self::get_ptr().write_volatile(v)
    }
    fn set_uart_reset(uart_id: i32, value: bool) {
        let mut flags = Self::from_bits(
            unsafe { Self::read() }
        ).unwrap();
        let bit = match uart_id {
            1 => Self::UART1_RESET,
            2 => Self::UART2_RESET,
            3 => Self::UART3_RESET,
            _ => { panic!("invalid uart id!"); }
        };
        flags.set(bit, value);
        unsafe { Self::write(flags.bits) }
    }
}
#[derive(Copy, Clone)]
pub enum SYSCTLClock {
    SYSCTLClockPLL0,
    SYSCTLClockPLL1,
    SYSCTLClockPLL2,
    SYSCTLClockCPU,
    SYSCTLClockSRAM0,
    SYSCTLClockSRAM1,
    SYSCTLClockAPB0,
    SYSCTLClockAPB1,
    SYSCTLClockAPB2,
    SYSCTLClockROM,
    SYSCTLClockDMA,
    SYSCTLClockAI,
    SYSCTLClockDVP,
    SYSCTLClockFFT,
    SYSCTLClockGPI0,
    SYSCTLClockSPI0,
    SYSCTlClockSPI1,
    SYSCTLClockSPI2,
    SYSCTLClockSPI3,
    SYSCTLClockI2S0,
    SYSCTLClockI2S1,
    SYSCTLClockI2S2,
    SYSCTLClockI2C0,
    SYSCTLClockI2C1,
    SYSCTLClockI2C2,
    SYSCTLClockUART1,
    SYSCTLClockUART2,
    SYSCTLClockUART3,
    SYSCTLClockAES,
    SYSCTLClockFPIOA,
    SYSCTLClockTimer0,
    SYSCTLClockTimer1,
    SYSCTLClockTimer2,
    SYSCTLClockWDT0,
    SYSCTLClockWDT1,
    SYSCTLClockSHA,
    SYSCTLClockOTP,
    SYSCTLClockRTC,
    SYSCTLClockACLK = 40,
    SYSCTLClockHCLK,
    SYSCTLClockIn0,
    SYSCTLClockMax,
}

pub fn sysctl_clock_enable(sysctl_clock: SYSCTLClock) {
    if sysctl_clock as i32 >= SYSCTLClock::SYSCTLClockMax as i32 {
        panic!("unknown sysctl clock type!");
    }
    sysctl_clock_bus_en(sysctl_clock, true);
    sysctl_clock_device_en(sysctl_clock, true);
}

fn sysctl_clock_bus_en(sysctl_clock: SYSCTLClock, enable: bool) -> i32 {
    if !enable {
        // we should disable related clock enable carefully
        return 0;
    }
    match sysctl_clock {
        /*
         * These peripheral devices are under APB0
         * GPIO, UART1, UART2, UART3, SPI_SLAVE, I2S0, I2S1,
         * I2S2, I2C0, I2C1, I2C2, FPIOA, SHA256, TIMER0,
         * TIMER1, TIMER2
         */
        SYSCTLClock::SYSCTLClockGPI0 |
        SYSCTLClock::SYSCTLClockSPI2 |
        SYSCTLClock::SYSCTLClockI2S0 |
        SYSCTLClock::SYSCTLClockI2S1 |
        SYSCTLClock::SYSCTLClockI2S2 |
        SYSCTLClock::SYSCTLClockI2C0 |
        SYSCTLClock::SYSCTLClockI2C1 |
        SYSCTLClock::SYSCTLClockI2C2 |
        SYSCTLClock::SYSCTLClockUART1 |
        SYSCTLClock::SYSCTLClockUART2 |
        SYSCTLClock::SYSCTLClockUART3 |
        SYSCTLClock::SYSCTLClockFPIOA |
        SYSCTLClock::SYSCTLClockTimer0 |
        SYSCTLClock::SYSCTLClockTimer1 |
        SYSCTLClock::SYSCTLClockTimer2 |
        SYSCTLClock::SYSCTLClockSHA => {
            SYSCTLClockEnableCent::set_apb_clock_enable(0, true);
        },
        /*
         * These peripheral devices are under APB1
         * WDT, AES, OTP, DVP, SYSCTL
         */
        SYSCTLClock::SYSCTLClockAES |
        SYSCTLClock::SYSCTLClockWDT0 |
        SYSCTLClock::SYSCTLClockWDT1 |
        SYSCTLClock::SYSCTLClockOTP |
        SYSCTLClock::SYSCTLClockRTC => {
            SYSCTLClockEnableCent::set_apb_clock_enable(1, true);
        },
        /*
         * These peripheral devices are under APB2
         * SPI0, SPI1
         */
        SYSCTLClock::SYSCTLClockSPI0 | SYSCTLClock::SYSCTlClockSPI1 => {
            SYSCTLClockEnableCent::set_apb_clock_enable(2, true);
        },
        _ => { panic!("sysctl clock type is not supported!"); }
    }
    0
}

fn sysctl_clock_device_en(sysctl_clock: SYSCTLClock, enable: bool) {
    match sysctl_clock {
        SYSCTLClock::SYSCTLClockUART1 |
        SYSCTLClock::SYSCTLClockUART2 |
        SYSCTLClock::SYSCTLClockUART3 => {
            SYSCTLClockEnablePeri::set_uart_clock_enable(
                sysctl_clock as i32 - SYSCTLClock::SYSCTLClockUART1 as i32 + 1,
                enable
            )
        }
        _ => {
            panic!("we cannot handle this now!")
        }
    }
}

#[derive(Copy, Clone)]
pub enum SYSCTLReset {
    SYSCTLResetSOC,
    SYSCTLResetROM,
    YSCTLResetDMA,
    SYSCTLResetAI,
    SYSCTLResetDVP,
    SYSCTLResetFFT,
    SYSCTLResetGPIO,
    SYSCTLResetSPI0,
    SYSCTLResetSPI1,
    SYSCTLResetSPI2,
    SYSCTLResetSPI3,
    SYSCTLResetI2S0,
    SYSCTLResetI2S1,
    SYSCTLResetI2S2,
    SYSCTLResetI2C0,
    SYSCTLResetI2C1,
    SYSCTLResetI2C2,
    SYSCTLResetUART1,
    SYSCTLResetUART2,
    SYSCTLResetUART3,
    SYSCTLResetAES,
    SYSCTLResetFPIOA,
    SYSCTLResetTimer0,
    SYSCTLResetTimer1,
    SYSCTLResetTimer2,
    SYSCTLResetWDT0,
    SYSCTLResetWDT1,
    SYSCTLResetSHA,
    SYSCTLResetRTC,
    SYSCTLResetMax
}

pub fn sysctl_reset(sysctl_reset: SYSCTLReset) {
    sysctl_reset_ctl(sysctl_reset, true);
    usleep(100);
    sysctl_reset_ctl(sysctl_reset, false);
}

pub fn sysctl_reset_ctl(sysctl_reset: SYSCTLReset, value: bool) {
    match sysctl_reset {
        SYSCTLReset::SYSCTLResetSOC => {
            SYSCTLSoftReset::set_soft_reset(value);
        },
        SYSCTLReset::SYSCTLResetUART1 |
        SYSCTLReset::SYSCTLResetUART2 |
        SYSCTLReset::SYSCTLResetUART3 => {
            SYSCTLPeriReset::set_uart_reset(sysctl_reset as i32 - SYSCTLReset::SYSCTLResetUART1 as i32 + 1, value);
        },
        _ => {
            panic!("unhandled sysctl_reset_ctl!");
        }
    }
}


